src/main/java/com/google/devtools/build/lib/worker/WorkRequestHandler.java

```diff
@@ -496,4 +500,100 @@ private void maybePerformGc() {
       }
     }
   }
+
+  /**
+   * Class that wraps the standard {@link System#in}, {@link System#out}, and {@link System#err}
+   * with our own ByteArrayOutputStream that allows {@link WorkRequestHandler} to safely capture
+   * outputs that can't be directly captured by the PrintStream associated with the work request.
+   *
+   * <p>This is most useful when integrating JVM tools that write exceptions and logs directly to
+   * {@link System#out} and {@link System#err} which corrupt the persistent worker.
+   *
+   * <p>WorkerIO implements {@link AutoCloseable} and will swap the original streams back into {@link
+   * System} once close has been called.
+   */
+  public static class WorkerIO implements AutoCloseable {
+
+    private final InputStream originalInputStream;
+    private final PrintStream originalOutputStream;
+    private final PrintStream originalErrorStream;
```

Comment:
Wrapping `System.err` isn't actually be necessary for a wrapper worker io.

Reply:
Correct. Also `System.in` is already handled properly. Only `System.out` needs to be wrapped.

Reply:
> Also System.in is already handled properly. Only System.out needs to be wrapped.

The motivation behind wrapping all of the system streams in a `WorkerIO` class is that there is a clear and consistent place to access the root system streams while setting up the work request handler. For `System.in` and `System.err` we can just return the original streams when calling `getOriginalInputStream`.

Reply:
> wrapping the inside of the WorkRequestCallback would be clearer

By this do you mean calling `capture` from within the implementation of `WorkRequestHandler`, or in the lambda that gets passed into the `WorkRequestHandler`? I like idea of this being hidden inside the `WorkRequestCallback` because we hide `WorkerIO#readCapturedAsUtf8String` entirely which is kind of an awkward API to need to expose.

> That change would also mean that the WorkerIO doesn't need to reset the 

Would each work request call `WorkerIO#capture` swapping in it's own buffer?

> A definite benefit of either approach is that stdout and stderr get merged into a single stream, that will be easier to understand.

Wrapping the `stderr` could be a bit error-prone because if there's ever an unrecoverable error that causes the worker to not send a work response, you just won't see an exception come out of the worker.

Reply:
> By this do you mean calling capture from within the implementation of WorkRequestHandler, or in the lambda that gets passed into the WorkRequestHandler?

From within the implementation, in `respondToRequest`. Putting it in there means the stdio redirection does not touch on the use of stdio for the server-worker protocol. 

> Would each work request call WorkerIO#capture swapping in it's own buffer?

Yes. Throw-away object is simplest object.

> Wrapping the stderr could be a bit error-prone because if there's ever an unrecoverable error that causes the worker to not send a work response, you just won't see an exception come out of the worker.

I believe that's already the case today, but I'll have to consider the various possible failures some more before I can opine on it.

How does your current solution deal with multiplex workers?

Reply:
> How does your current solution deal with multiplex workers?

@larsrc-google I took another pass at building this into the `WorkRequestHandler` so that it's a bit more transparent, and with the multiplex worker cases in mind. The new approach uses the existing `WorkerIO` implementation, but wraps the system streams when you call `WorkRequestHandler#processRequests`. 

The `WorkerIO` instance gets passed down into `respondToRequest` so that there is a single source of truth for the wrapped streams, and so that we don't have to deal with multiple `WorkerIO` instances being created inside of `respondToRequest` which doesn't really make sense in a multiplex-world where `respondToRequest` may get called concurrently.

Reply:
Oh, I've been so focused on the implementation details that I forgot the larger problem: Redirecting the stderr messages into the `output` field of the response changes semantics. The `output` field is for shorter messages to be displayed to the user, `stderr` goes into the logs. See https://docs.bazel.build/versions/main/creating-workers.html#work-responses. It's known that multiplex workers only have one log between them, but the messages shown to the user should not interleave.

And of course, since the multiplex threads still share one `System.err` instance regardless of what we set, we can't possibly disentangle that. But we can still fix the main issue of stdout getting corrupted by setting it to the same stream as stderr.

Reply:
Just to make sure i'm understanding next steps here - You'd prefer that we avoid capturing the tool output streams and instead just swap the `System.err` in for `System.out` to not pollute the work response output?

Reply:
Yes. Having System.err and System.out go to the same place (the logs) is good. Changing the work response output would cause a lot of noise.

Reply:
Pushed the changes for swapping `System.err` in for `System.out` and updated the tests.